lib/streamlit/elements/doc_string.py

```diff
@@ -86,66 +125,416 @@ def _marshall(doc_string_proto: DocStringProto, obj: Any) -> None:
 
+def _get_name(obj):
+    # Try to get the fully-qualified name of the object.
+    # For example:
+    #   st.help(bar.Baz(123))
+    #
+    #   The name is bar.Baz
+    name = getattr(obj, "__qualname__", None)
+    if name:
+        return name
+
+    # Try to get the name of the object.
+    # For example:
+    #   st.help(bar.Baz(123))
+    #
+    #   The name is Baz
+    name = getattr(obj, "__name__", None)
+    return name
+
+
+def _get_module(obj):
+    return getattr(obj, "__module__", None)
+
+
+def _get_signature(obj):
+    if not inspect.isclass(obj) and not callable(obj):
+        return None
+
+    sig = ""
+
+    # TODO: Can we replace below with this?
+    # with contextlib.suppress(ValueError):
+    #     sig = str(inspect.signature(obj))
+
+    try:
+        sig = str(inspect.signature(obj))
+    except ValueError:
+        sig = "(...)"
+    except TypeError:
+        return None
 
-def _get_signature(f):
     is_delta_gen = False
     with contextlib.suppress(AttributeError):
-        is_delta_gen = f.__module__ == "streamlit.delta_generator"
+        is_delta_gen = obj.__module__ == "streamlit.delta_generator"
         # Functions such as numpy.minimum don't have a __module__ attribute,
         # since we're only using it to check if its a DeltaGenerator, its ok
         # to continue
 
-    sig = ""
-
-    with contextlib.suppress(ValueError):
-        sig = str(inspect.signature(f))
     if is_delta_gen:
         for prefix in CONFUSING_STREAMLIT_SIG_PREFIXES:
             if sig.startswith(prefix):
                 sig = sig.replace(prefix, "(")
                 break
 
     return sig
+
+
+class UNSET:
+    pass
+
+
+def _get_docstring(obj):
+    doc_string = inspect.getdoc(obj)
+
+    # Sometimes an object has no docstring, but the object's type does.
+    # If that's the case here, use the type's docstring.
+    # For objects where type is "type" we do not print the docs (e.g. int).
+    # We also do not print the docs for functions and methods if the docstring is empty.
+    if doc_string is None:
+        obj_type = type(obj)
+
+        if (
+            obj_type is not type
+            and obj_type is not types.ModuleType
+            and not inspect.isfunction(obj)
+            and not inspect.ismethod(obj)
+        ):
+            doc_string = inspect.getdoc(obj_type)
+
+    if doc_string:
+        return doc_string.strip()
+
+    return None
+
+
+_NEWLINES = re.compile(r"[\n\r]+")
+
+
+def _get_variable_name():
+    """Try to get the name of the variable in the current line, as set by the user.
+
+    For example:
+    foo = bar.Baz(123)
+    st.help(foo)
+
+    The name is "foo"
+    """
+    scriptrunner_frame = _get_scriptrunner_frame()
+
+    if scriptrunner_frame is None:
+        # If there's no ScriptRunner frame, something weird is going on. This should never happen
+        # but let's bail out nicely just in case there's some valid edge case where this is OK.
```

Comment:
I think one edge case where this may happen is if a user calls a script with `st.*` calls directly via `python my_script.py`, in which case we want every `st.*` call to be more or less a no-op and just want to print a message to the user telling them to use `streamlit run`

May be worth  mentioning this as the one place we expect this in the comment

Reply:
Great catch. Will add.